dvc/repo/ls.py

```diff
@@ -0,0 +1,108 @@
+import os
+
+from dvc.exceptions import DvcException, NotDvcRepoError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    if os.path.exists(url):
+        if rev:
+            raise DvcException("Local repo doesn't support revisions")
+        return _ls_local_repo(url, target, recursive, outputs_only)
+
+    return _ls_remote_git_repo(url, target, rev, recursive, outputs_only)
+
+
+def _ls_files_local_repo(path, target=None, recursive=None):
+    from dvc.scm.tree import WorkingTree
+    from dvc.ignore import CleanTree
+
+    def prettify(prefix, filename):
+        if not prefix == os.curdir:
+            filename = os.path.join(prefix, filename)
+        return filename
+
+    files = []
+    target_path = os.path.join(path, target) if target else path
+    tree = CleanTree(WorkingTree(target_path))
+
+    try:
+        target_path = os.path.join(path, target) if target else path
+        if not os.path.exists(target_path):
+            raise FileNotFoundError("{} doesn't exist".format(target_path))
+
+        if target:
+            files.append(prettify(os.curdir, target))
+            if not recursive:
+                return files
+
+        for dirpath, dirnames, filenames in tree.walk(target_path):
+            prefix = os.path.relpath(dirpath, path)
+            files.extend(map(lambda f: prettify(prefix, f), dirnames))
+            files.extend(map(lambda f: prettify(prefix, f), filenames))
+            if not recursive:
+                break
+    except (NotADirectoryError, FileNotFoundError):
+        pass
+
+    return files
+
+
+def _ls_outs_local_repo(path, target=None, recursive=None):
+    from dvc.repo import Repo
+    from dvc.path_info import PathInfo
+
+    try:
+        repo = Repo(path)
+    except NotDvcRepoError:
+        return []
+
+    outs = [out for stage in repo.stages for out in stage.outs]
+    target_path = os.path.join(path, target) if target else path
+    path_info = PathInfo(os.path.abspath(target_path))
+
+    def prettify(out):
+        relpath = out.path_info.relpath(path)
+        parts = list(relpath.parts)
+        segments = [parts.pop(0)]
+        if not recursive:
+            return segments
+
+        while parts:
+            segments.append(os.path.join(segments[-1], parts.pop(0)))
+        return segments
+
+    files = set()
+    for out in outs:
+        if out.path_info.isin_or_eq(path_info):
+            if not recursive and target:
+                files.add(target)
+                break
+
+            files.update(prettify(out))
+
+    repo.close()
```

Comment:
Here is where you are forgetting to put it into the `finally` block or something :) Hopefully the refactoring that I've mentioned above will handle this.

Reply:
I haven't got you point, unfortunately
I close repo at the end of the work and use `try`/`except` above trying to open repo

Reply:
If any error happens between `Repo()` and this `close()` you will leak resources. Better to try to organise your code in such a way to not leak anything even if you get some error that you don't expect. I think it will be solved by abstracting above, so that it is similar to `with external_repo` that closes your repo instance automatically when you are exiting the context, with error or without.